tests/30rooms/06invite.pl

```diff
@@ -107,3 +111,424 @@
+         my ( $f ) = @_;
+         $f->then( sub {
+            Future->done( @_ );
+         }, sub {
+            my ( undef, $name, $response ) = @_;
+
+            Future->fail( @_ );
+         });
+      }
+      : \&main::expect_http_4xx;
+
+   Future->needs_all(
+      stub_is_lookup( $invitee_email, undef, $await_http_request ),
+
+      stub_is_token_generation( $token, $encoded_public_key, $await_http_request ),
+
+      @$futures,
+
+      matrix_create_room( $inviter, visibility => "private" )
+      ->then(sub {
+         my ( $room_id ) = @_;
+         do_3pid_invite( $inviter, $room_id, $id_server, $invitee_email )
+         ->then( sub {
+            $join_sub->( $token, $public_key, $signature, $room_id )
+            ->followed_by($response_verifier)
+            ->then( sub {
+               matrix_get_room_state( $inviter, $room_id,
+                  type => "m.room.member",
+                  state_key => $invitee->user_id,
+               )
+            })->followed_by(assert_membership( $inviter, $expect_join_success ? "join" : undef ) );
+         })
+      }),
+   );
+}
+
+sub assert_membership {
+   my ( $user, $expected_membership ) = @_;
+
+   my $verifier = defined $expected_membership
+      ? sub {
+         my ( $f ) = @_;
+
+         $f->then( sub {
+            my ( $body ) = @_;
+
+            log_if_fail "Body", $body;
+            $body->{membership} eq $expected_membership or
+               die "Expected invited user membership to be '$expected_membership' but was '".$body->membership."'";
+
+            Future->done( 1 );
+         } )
+      }
+      : \&main::expect_http_404;
+};
+
+sub stub_is_lookup {
+   my ( $email, $mxid, $await_http_request ) = @_;
+
+   $await_http_request->("/_matrix/identity/api/v1/lookup", sub {
+      my ( $req ) = @_;
+      return unless $req->query_param("medium") eq "email";
+      return unless $req->query_param("address") eq $email;
+      return 1;
+   })->then( sub {
+      my ( $request ) = @_;
+      $request->respond_json(defined($mxid) ? {medium => "email", address => $email, mxid => $mxid} : {});
+      Future->done( 1 );
+   })
+};
+
+sub stub_is_token_generation {
+   my ( $token, $encoded_public_key, $await_http_request ) = @_;
+
+   $await_http_request->( "/_matrix/identity/api/v1/nonce-it-up", sub {
+      my ( $req ) = @_;
+      # TODO: Parse body
+      return 1;
+   })->then( sub {
+      my ( $request ) = @_;
+      $request->respond_json( {
+            token => $token,
+            public_key => $encoded_public_key,
+         } );
+      Future->done( 1 );
+   })
+};
+
+sub stub_is_key_validation {
+   my ( $validity, $await_http_request, $user_agent_substring ) = @_;
+
+   $await_http_request->( "/_matrix/identity/api/v1/pubkey/isvalid", sub {
+      my ( $req ) = @_;
+      !defined $user_agent_substring and return 1;
+      my $user_agent = $req->header( "User-Agent" );
+      defined $user_agent and index( $user_agent, $user_agent_substring ) ge 0
```

Comment:
A more idiomatic substring test is probably to use a regexp match

    defined $user_agent and $user_agent =~ m/\Q$wanted_user_agent/ or
       return 0;